libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+	"github.com/ava-labs/libevm/log"
+	"github.com/ava-labs/libevm/trie/trienode"
+	"github.com/ava-labs/libevm/trie/triestate"
+	"github.com/ava-labs/libevm/triedb"
+	"github.com/ava-labs/libevm/triedb/database"
+)
+
+const (
+	dbFileName  = "firewood.db"
+	logFileName = "firewood.log"
+)
+
+var _ triedb.DBOverride = (*Database)(nil)
+
+type Database struct {
+	// The underlying Firewood database, used for storing proposals and revisions.
+	// This must be exported so other packages (e.g. state sync) can access firewood-specific methods.
+	Firewood *ffi.Database
+
+	// proposalLock protects the proposal map and tree during updates.
+	proposalLock sync.RWMutex
+	// proposalMap provides O(1) access by state root to all proposals stored in the proposalTree
+	proposalMap map[common.Hash][]*proposal
+	// The proposal tree tracks the structure of the current proposals, and which proposals are children of which.
+	// This is used to ensure that we can dereference proposals correctly and commit the correct ones
+	// in the case of duplicate state roots.
+	// The root of the tree is stored here, and represents the top-most layer on disk.
+	proposalTree *proposal
+
+	// possibleProposals temporarily holds proposals created during a trie update.
+	// This is cleared after the update is complete and the proposals have been sent to the database.
+	possibleProposals []*proposal
+}
+
+type Config struct {
+	ChainDir             string
+	CleanCacheSize       int  // Size of the clean cache in bytes
+	FreeListCacheEntries uint // Number of free list entries to cache
+	Revisions            uint
+	ReadCacheStrategy    ffi.CacheStrategy
+}
+
+// Note that `FilePath` is not specified, and must always be set by the user.
+var Defaults = Config{
+	CleanCacheSize:       1024 * 1024, // 1MB
+	FreeListCacheEntries: 40_000,
+	Revisions:            100,
+	ReadCacheStrategy:    ffi.CacheAllReads,
+}
+
+func (c Config) BackendConstructor(ethdb.Database) triedb.DBOverride {
+	db, err := New(c)
+	if err != nil {
+		log.Crit("firewood: error creating database", "error", err)
+	}
+	return db
+}
+
+// New creates a new Firewood database with the given disk database and configuration.
+// Any error during creation will cause the program to exit.
+func New(config Config) (*Database, error) {
+	if err := validatePath(config.ChainDir); err != nil {
+		return nil, err
+	}
+
+	// Start the logs prior to opening the database
+	logPath := filepath.Join(config.ChainDir, logFileName)
+	if err := ffi.StartLogs(&ffi.LogConfig{Path: logPath}); err != nil {
+		// This shouldn't be a fatal error, as this can only be called once per thread.
+		// Specifically, this will return an error in unit tests.
+		log.Warn("firewood: error starting logs", "error", err)
+	}
+
+	dbPath := filepath.Join(config.ChainDir, dbFileName)
+	fw, err := ffi.New(dbPath, &ffi.Config{
+		NodeCacheEntries:     uint(config.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
```

Comment:
```suggestion
		NodeCacheEntries:     uint(config.CleanCacheSize) / 256, // TODO(alarso16): estimate 256 bytes per node
```

FWIW, a to-do label isn't necessarily so that person does the work, but as a signal of who to ask for more information. Re the message, is it saying that 256 bytes is the current estimate, or that we need to estimate?

Reply:
That is the current estimate, which is probably inaccurate